design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,542 @@
+---
+
+# XDN11 - Error handling
+
+- Author: Manodasan Wignarajah (mawign@microsoft.com)
+- Status: Draft
+
+## Abstract
+
+This document describes the error handling story for xlang and what it will support.
+
+## Overview
+
+As described in XDN01, xlang is a system that allows a component written in a supported language
+to be callable from any other supported language. Upon calling a component, errors can occur and
+can play a different role based on the audience and the situation.
+
+Errors are used by component implementers to indicate both expected and unexpected issues. Expected
+issues are situations which a component implementer foresees happening in a normal scenario such as
+not being able to open a file and wants to communicate it to the consumer. In these cases, xlang
+recommends the use of the Try Pattern to communicate these issues. Unexpected issues are situations
+that are not foreseen in a normal scenario such as a memory allocation error. In these cases, a
+component implementer may use custom logging to log the issue to help them with diagnosing it.
+They can also propagate the error back to the consumer which xlang will translate from the
+component's language to the consumer's language.
+
+Errors are also used by projection implementers to indicate an issue while translating a call from
+one language to the other. These errors are also translated by xlang to the consumer's language.
+
+Errors from components and projections are received by the consumer and either handled, propagated,
+or result in a failure. If it results in a failure, the failure might be debugged by a developer,
+logged by the application, or captured in a dump. To help facilitate this in a way that the developer
+has information available to them to diagnose these translated errors, xlang producing projections
+retrieve the error context captured by the component's language and store it as part of the
+xlang error info when it is originated. This xlang error info is then stored with the consumer's
+language error by the consuming projection and is available to be logged and captured. If the consumer
+propagates the error, the error may end up traveling to another xlang projected component before
+reaching the failure point. To facilitate diagnosing these failures, xlang allows any producing
+projections which the error propagates through to append additional error context captured by that
+language to the existing xlang error info. The information together show how the error propagated
+and resulted in a failure.
+
+![Error origination diagram](XDN11_origination_diagram.png)
+Diagram highlighting error handling of a C# app calling a Python API which calls a Java API that results in an error.
+
+## Guidance for Projection Authors
+
+- **DO** translate language errors to equivalent xlang errors and xlang errors to equivalent natural
+language errors at the projection boundary.
+- **DO NOT** let language errors propagate past the projection boundary.
```

Comment:
be sure to emphasize that it's exceptions specifically that must not pass the boundary

Reply:
Addressed.